lib/alexa_request_verifier/verifier.rb

```diff
@@ -4,6 +4,121 @@
 module AlexaRequestVerifier
   # A namespace for all of our verifiers to live under
   # @since 0.1
-  module Verifier
+  class Verifier
+    attr_accessor :configuration
+
+    # Create a new AlexaRequestVerifier::Verifier object
+    #
+    # @yield the configuration block
+    # @yieldparam config [AlexaRequestVerifier::Configuration] the configuration object
+    def initialize
+      @configuration = AlexaRequestVerifier::Configuration.new
+
+      yield @configuration if block_given?
+    end
+
+    # Validate a request object from Sinatra.
+    # Raise an error if it is not valid.
+    #
+    # @param [Sinatra::Request] request a Sinatra HTTP Request
+    #
+    # @raise [AlexaRequestVerifier::InvalidCertificateURIError]
+    #   there was a problem validating the certificate URI from your request
+    #
+    # @return [nil] will always return nil
+    def valid!(request)
+      signature_certificate_url = request.env['HTTP_SIGNATURECERTCHAINURL']
+
+      AlexaRequestVerifier::Verifier::CertificateURIVerifier.valid!(signature_certificate_url) if @configuration.verify_uri?
+
+      raw_body = request.body.read
+      request.body.rewind
+
+      check_that_request_is_timely(raw_body) if @configuration.verify_timeliness?
+
+      check_that_request_is_valid(signature_certificate_url, request, raw_body)
+
+      true
+    end
+
+    # Validate a request object from Sinatra.
+    # Return a boolean.
+    #
+    # @param [Sinatra::Request] request a Sinatra HTTP Request
+    # @return [Boolean] is the request valid?
+    def valid?(request)
+      begin
+        valid!(request)
+      rescue AlexaRequestVerifier::BaseError => e
+        puts e
+
+        return false
+      end
+
+      true
+    end
+
+    # Used to configure AlexaRequestVerifier.
+    #
+    # @example
+    #    AlexaRequestVerifier.configure do |c|
+    #      c.some_config_option = true
+    #    end
+    #
+    # @yield the configuration block
+    # @yieldparam config [AlexaRequestVerifier::Configuration] the configuration object
+    def configure
+      yield @configuration
+    end
+
+    private
+
+    # Prevent replays of requests by checking that they are timely.
+    #
+    # @param [String] raw_body the raw body of our https request
+    # @raise [AlexaRequestVerifier::InvalidRequestError] raised when the timestamp is not timely, or is not set
+    def check_that_request_is_timely(raw_body)
+      request_json = JSON.parse(raw_body)
+
+      raise AlexaRequestVerifier::InvalidRequestError, 'Timestamp field not present in request' if request_json.fetch('request', {}).fetch('timestamp', nil).nil?
+
+      raise AlexaRequestVerifier::InvalidRequestError, 'Request is from more than 150 seconds ago' unless Time.parse(request_json['request']['timestamp'].to_s) >= (Time.now - REQUEST_THRESHOLD)
```

Comment:
Should `REQUEST_THRESHOLD` be configurable?

Reply:
Please see above